circuits/cpp/src/aztec3/circuits/kernel/public/common.hpp

```diff
@@ -0,0 +1,178 @@
+#include <aztec3/utils/array.hpp>
+#include <aztec3/circuits/hash.hpp>
+
+using NT = aztec3::utils::types::NativeTypes;
+using aztec3::circuits::abis::KernelCircuitPublicInputs;
+using aztec3::circuits::abis::PublicDataWrite;
+using aztec3::circuits::abis::StateRead;
+using aztec3::circuits::abis::StateTransition;
+using aztec3::circuits::abis::public_kernel::PublicKernelInputs;
+using aztec3::circuits::abis::public_kernel::PublicKernelInputsNoPreviousKernel;
+using DummyComposer = aztec3::utils::DummyComposer;
+using aztec3::circuits::compute_public_data_tree_index;
+using aztec3::circuits::compute_public_data_tree_value;
+using aztec3::circuits::root_from_sibling_path;
+using aztec3::utils::array_length;
+using aztec3::utils::array_pop;
+using aztec3::utils::array_push;
+using aztec3::utils::push_array_to_array;
+
+namespace aztec3::circuits::kernel::public_kernel {
+
+template <size_t SIZE>
+void check_membership(DummyComposer& composer,
+                      NT::fr const& value,
+                      NT::fr const& index,
+                      std::array<NT::fr, SIZE> const& sibling_path,
+                      NT::fr const& root,
+                      std::string const& msg)
+{
+    const auto calculated_root = root_from_sibling_path<NT>(value, index, sibling_path);
+    composer.do_assert(calculated_root == root, std::string("Membership check failed ") + msg);
+}
+
+template <typename KernelInput>
+void validate_state_reads(DummyComposer& composer, KernelInput const& public_kernel_inputs)
+{
+    // Validates that state reads correspond to the provided membership witnesses
+    const auto& reads = public_kernel_inputs.public_call.public_call_data.call_stack_item.public_inputs.state_reads;
+    const auto& contract_address = public_kernel_inputs.public_call.public_call_data.call_stack_item.contract_address;
+    for (size_t i = 0; i < STATE_READS_LENGTH; ++i) {
+        const auto& state_read = reads[i];
+        if (state_read.is_empty()) {
+            continue;
+        }
+        const auto& sibling_path = public_kernel_inputs.public_call.state_reads_sibling_paths[i].sibling_path;
+        const NT::fr leaf_value = compute_public_data_tree_value<NT>(state_read.current_value);
+        const NT::fr leaf_index = compute_public_data_tree_index<NT>(contract_address, state_read.storage_slot);
+        const std::string msg = format("validate_state_reads, index ", i, " leaf value ", leaf_value);
+        check_membership(composer,
+                         leaf_value,
+                         leaf_index,
+                         sibling_path,
+                         public_kernel_inputs.public_call.public_data_tree_root,
+                         msg);
+    }
+};
+
+template <typename KernelInput>
+void validate_state_transitions(DummyComposer& composer, KernelInput const& public_kernel_inputs)
+{
+    // Validates that the old value of state transitions correspond to the provided membership witnesses
```

Comment:
Again, code reads pretty clearly, but it might be good to to elaborate just a bit in the top comment. But honestly looks pretty good so it's not super important.